### PR TITLE
Fixes #2022 Add common interface to indicate a building block contain…

### LINKED ISSUE #2022: Add common interface to indicate a building block contains initial conditions
This relates to https://github.com/Open-Systems-Pharmacology/MoBi/issues/855

### PR DESCRIPTION
**Pull request by rwmcintosh**
…s initial conditions

### CODE REVIEW

File: src/OSPSuite.Core/Comparison/ExpressionProfileDiffBuilder.cs [hunk 1]
```diff
@@ -12,7 +13,7 @@ public override void Compare(IComparison<ExpressionProfileBuildingBlock> compari
       {
          base.Compare(comparison);
          CompareValues(x => x.Type, x => x.Type, comparison);
-         _enumerableComparer.CompareEnumerables(comparison, x => x.InitialConditions, x => x.Path);
```
Comment (msevestre) [inline src/OSPSuite.Core/Comparison/ExpressionProfileDiffBuilder.cs:15]: I am not sure why this was removed? It makes sense for me to have a property InitialCoditions get at the ExpressionProfile level. Using the expression profile as an IEnumerable InitialCondition is not *intuitive*

File: src/OSPSuite.Core/Domain/Builder/ExpressionProfileBuildingBlock.cs [hunk 2]
```diff
@@ -76,7 +59,25 @@ public override void UpdatePropertiesFrom(IUpdatable source, ICloneManager clone
          Name = sourceExpressionProfile.Name;
 
          _initialConditions.Clear();
-         sourceExpressionProfile.InitialConditions.Each(initialCondition => AddInitialCondition(cloneManager.Clone(initialCondition)));
+         sourceExpressionProfile.Each<InitialCondition>(initialCondition => Add(cloneManager.Clone(initialCondition)));
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Builder/ExpressionProfileBuildingBlock.cs:62]: Same here. Each with an <> is very unexpected code. I think I want the old code back

File: src/OSPSuite.Core/Domain/Builder/SimulationBuilder.cs [hunk 3]
```diff
@@ -93,7 +93,7 @@ private void performMerge()
          _molecules.AddRange(allBuilder(x => x.Molecules));
          _parameterValues.AddRange(allStartValueBuilder(x => x.SelectedParameterValues));
          _initialConditions.AddRange(allStartValueBuilder(x => x.SelectedInitialConditions)
-            .Concat(_simulationConfiguration.ExpressionProfiles.SelectMany(x => x.InitialConditions)));
+            .Concat(_simulationConfiguration.ExpressionProfiles.SelectMany<ExpressionProfileBuildingBlock, InitialCondition>(x => x)));
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Builder/SimulationBuilder.cs:96]: lol ok this is even more extreme. This code becomes unreadable

File: src/OSPSuite.Core/Serialization/Xml/BuildingBlockXmlSerializer.cs [hunk 4]
```diff
@@ -118,7 +118,7 @@ public override void PerformMapping()
       {
          base.PerformMapping();
          Map(x => x.Type);
-         MapEnumerable(x => x.InitialConditions, x => x.AddInitialCondition);
+         MapEnumerable<InitialCondition>(bb => bb, bb => bb.Add).WithMappingName(Constants.Serialization.INITIAL_CONDITIONS);
```
Comment (msevestre) [inline src/OSPSuite.Core/Serialization/Xml/BuildingBlockXmlSerializer.cs:121]: Please revert this also

**Review by msevestre — CHANGES_REQUESTED** [review]
I do not like this change at all. I think adding the IBuidlignBlcok is one thing that we can agree upon, but removing the accessor to InitialCondition is going to be a harder sell for me and this was *not* what you explained. Each instance where it was removed results in code that is not idiomatic.
I am in favor of adding the InitialConditions accessor back, undoing all changes where it was removed in favor of some weird generics and go from there

**Comment by rwmcintosh** [comment]
> I do not like this change at all. I think adding the IBuidlignBlock is one thing that we can agree upon, but removing the accessor to InitialCondition is going to be a harder sell for me and this was _not_ what you explained. Each instance where it was removed results in code that is not idiomatic. I am in favor of adding the InitialConditions accessor back, undoing all changes where it was removed in favor of some weird generics and go from there

Ok, that makes sense. That means that in places where you have the most derived version ```ExpressionProfileBuildingBlock``` you can use the accessor to get the initial conditions. You are not forced to get the enumerator of the building block, sometimes using weird generics.

**Review by msevestre — APPROVED** [review]

**Pull request merged**